seq2simulate/hypermutation/hypermutate.py

```diff
@@ -0,0 +1,167 @@
+        """
+        sequence = sequence.upper()
+        if not all(nucleotide in 'ATGC-' for nucleotide in set(sequence)):
+            raise ValueError('Sequences may only contain A, T, G, and/or C ({})'.format(list(set(sequence))))
+
+        self.sequence = sequence
+        self.sequence_motif_dict = None
+        if motif_probabilities is None:
+            self.motif_probabilities = mutation_probabilities.KijakProbabilities()
+        else:
+            self.motif_probabilities = motif_probabilities
+
+    def _get_motif_indices(self, motif):
+        """
+        Find indices of substring motif in sequence.
+    
+        Args:
+            motif: a substring
+    
+        Returns: a list of left indices
+        """
+        indices = [m.start() for m in re.finditer(re.escape(motif), self.sequence)]
+        return indices
+
+    def index_all_motifs(self):
+        """
+        Returns a dictionary of all motif indices in the sequence.
+        """
+
+        motif_dict = {}
+        for motif in self.motif_probabilities.motifs:
+            motif_indices = self._get_motif_indices(motif) 
+            if len(motif_indices) > 0:
+                motif_dict[motif] = motif_indices
+
+        self.sequence_motif_dict = motif_dict
+
+   
+    @property 
+    def _num_motifs(self):
+        """
+        Returns number of substrate motifs currently present in sequence.
+        """
+        try:
+            _num_motifs = sum([len(j) for i, j in self.sequence_motif_dict.iteritems()])
+        except AttributeError:
+            _num_motifs = 0
+        return _num_motifs
+
+    @property
+    def _motif_probs(self):
+        """
+        A shortened rescaled dictionary of motif mutation probabilities
+        including only the present motifs.
+
+        """
+        try:
+            remaining_probabilities_sum = sum([self.motif_probabilities.motifs[i] for i in self.sequence_motif_dict])
+            remaining_probabilities = {i: self.motif_probabilities.motifs[i] / remaining_probabilities_sum for i in self.sequence_motif_dict}
+        except AttributeError:
+            remaining_probabilities = {}
+        return remaining_probabilities
+            
+
+    @property 
+    def _motif_probs_cum(self):    
+        return self._get_cumulative_p(self._motif_probs)
+
+
+    def mutate_sequence(self, n):
+        """
+        Attempt to perform n mutations of self.sequence or until no more substrate motifs exist.
+
+        Args:
+            n: int
+        """
+
+        if self.sequence_motif_dict is None:
+            self.index_all_motifs()
+        if n > self._num_motifs:
+            n = self._num_motifs
+            #raise ValueError('n must be less than or equal to _num_motifs ({})'.format(self._num_motifs))
+            
+        mutations = 0
+        while mutations < n:
+            self.sequence = self._mutate(self.sequence)
+            self.index_all_motifs()
```

Comment:
Mutating without reindexing motifs means your object is in an inconsistent state: this should be moved inside _mutate.